### PR TITLE
Add API for mods

### PR DESCRIPTION
**Pull request by kremnev8**
This allows other mods to easily support Nebula without using reflection:
- Mods can reference classes such as `LocalPlayer`, `FactoryManager`, `SimulatedWorld` and `BinaryUtils`
- Allows registering and sending packets
- Checking mods version and syncing settings to clients

In the process I had to make some classes singleton, so there is a lot of small changes everywhere. I also removed hard-coded Galactic Scale compatibility code, however Galactic Scale can now use the API with minimal changes to its code.
I also have successfully tested the API against some of my WIP mods and Galactic Scale 2.
We will need to release this as separate package on Thunderstore to make releasing new version of the API possible.

### CODE REVIEW

File: NebulaModel/Packets/PacketProcessor.cs
Comment (Mushroom) [inline NebulaModel/Packets/PacketProcessor.cs:10]: Which "ProcessPacket" is this calling? I think prefixing this with the relevent class would be better here (ie, `base.` or `this.`) to better disambiguate

File: NebulaNetwork/PacketProcessors/Factory/Entity/BuildEntityRequestProcessor.cs
Comment (Mushroom) [inline NebulaNetwork/PacketProcessors/Factory/Entity/BuildEntityRequestProcessor.cs:17]: Why is this commented out?

File: NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs
Comment (Mushroom) [inline NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs:83]: What is the reason for removing these spaces?

File: NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs
Comment (Mushroom) [inline NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs:87]: Commented out code - please remove if not required

File: NebulaPatcher/NebulaPlugin.cs
Comment (Mushroom) [inline NebulaPatcher/NebulaPlugin.cs:88]: Why is this always fixed to true?

File: NebulaWorld/Factory/FactoryManager.cs
Comment (Mushroom) [inline NebulaWorld/Factory/FactoryManager.cs:23]: We do not use underscores

**Review by Mushroom — CHANGES_REQUESTED** [review]

**Review by hubastard — CHANGES_REQUESTED** [review]

File: NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs
Comment (kremnev8) [reply NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs:83]: This is string interpolation, spaces here mean nothing. It just my IDE formatting the file.

File: NebulaAPI/Data/IModData.cs
Comment (kremnev8) [reply NebulaAPI/Data/IModData.cs:1]: I have no idea. It was autogenerated by IDE.

File: NebulaNetwork/PacketProcessors/Factory/Entity/BuildEntityRequestProcessor.cs
Comment (kremnev8) [reply NebulaNetwork/PacketProcessors/Factory/Entity/BuildEntityRequestProcessor.cs:17]: I was commenting these lines during testing, i think i forgot about this one.

File: NebulaPatcher/NebulaPlugin.cs
Comment (kremnev8) [reply NebulaPatcher/NebulaPlugin.cs:88]: That means we want to check version of mod. If its set to false we only check if mod exists on both sides.

File: NebulaModel/Packets/PacketProcessor.cs
Comment (kremnev8) [reply NebulaModel/Packets/PacketProcessor.cs:10]: I think cast to NebulaConnection disambiguates it. Plus both really belong to this.

File: NebulaNetwork/MultiplayerClientSession.cs
Comment (kremnev8) [reply NebulaNetwork/MultiplayerClientSession.cs:87]: I have added this.

File: NebulaAPI/World/IFactoryManager.cs
Comment (kremnev8) [reply NebulaAPI/World/IFactoryManager.cs]: I have moved these constants to `NebulaModAPI` class. Is that better?

File: NebulaPatcher/NebulaPlugin.cs
Comment (Shad0wlife) [reply NebulaPatcher/NebulaPlugin.cs:88]: Typo: Verson => Version (missing i)

Comment (kremnev8) [reply NebulaPatcher/NebulaPlugin.cs:88]: Oops.

**Comment by PhantomGamers** [comment]
In order to publish this as a package on Thunderstore we will need a README.md file describing the purpose of the API along with instructions (or a link to instructions) on how it should be used, as well as an icon.png file (I guess I can just throw "API" onto the image we have for the base mod).

@kremnev8 could you do the instructions, ideally have a small example on how another mod would utilize it

**Comment by PhantomGamers** [comment]
Maybe just something like this for the icon?
thunderstore_icon_api

**Comment by kremnev8** [comment]
I have added documentation. I think it should be on Nebula Wiki page. Unfortunately i can't add a wiki page with this PR, so upon merging this [page](https://github.com/kremnev8/nebula/wiki/Nebula-mod-API) needs to be added to Nebula Wiki pages

File: NebulaPatcher/NebulaPlugin.cs
Comment (kremnev8) [reply NebulaPatcher/NebulaPlugin.cs]: I have decided to entirely remove this option. On a second look it just does not make any sense

**Comment by PhantomGamers** [comment]
So I think the holdups before this gets merged is:

- @hubastard mentioned wanting to refactor the multiplayer sessions and singletons before merging this
image
(would this be #228 ?)

- @Mushroom needs to review the changes made since his comments

- and I need to push the code to automate releases of the api

**Comment by kremnev8** [comment]
@PhantomGamers In my last commit I think I again accidentally changed websocket-sharp to a different commit. I can't figure where its set to correct commit, can you fix it for me?

**Comment by kremnev8** [comment]
As far as i can tell this PR is ready for merging. Only waiting for @Mushroom review.

**Comment by PhantomGamers** [comment]
This looks good to me but I'll hold off on actually approving it (which would trigger a CI build) in case someone has an objection to some part of the API

**Review by PhantomGamers — COMMENTED** [review]
just sneaking this fix into this PR 😂

**Review by hubastard — APPROVED** [review]

**Review by PhantomGamers — APPROVED** [review]

**Pull request merged**
